File: packages/protocol/src/agent-lifecycle.ts
/** Room-wide roster rules. See fixtures/agent-lifecycle.json for cross-client conformance. */
export const agentLifecyclePolicy = Object.freeze({
  reconnectGraceMs: 30_000,
  receptionFreshnessMs: 15_000,
  archiveAfterDays: 7,
  recentOfflineLimit: 100,
  pageSize: 100,
});

export type AgentConnection = 'online' | 'reconnecting' | 'offline';
export type AgentReceptionState = 'waiting' | 'on_resume' | 'unknown' | 'unavailable';
export type AgentArchiveReason = 'expired' | 'capacity';
export type AgentPresenceEvidence = {
  readonly agentId: string;
  readonly reportedStatus: string;
  readonly leaseExpiresAtUnixMs: number;
  readonly lastActiveAtUnixMs: number;
  readonly lastPolledAtUnixMs?: number;
  readonly listeningUntilUnixMs?: number | null;
};
export type AgentLifecycle = {
  readonly offlineSinceUnixMs: number | null;
  readonly archived: boolean;
  readonly archiveReason: AgentArchiveReason | null;
} & (
  | {
      readonly connection: 'online';
      readonly reception: Exclude<AgentReceptionState, 'unavailable'>;
    }
  | { readonly connection: 'reconnecting' | 'offline'; readonly reception: 'unavailable' }
);

export function agentConnection(evidence: AgentPresenceEvidence, now: number): AgentConnection {
  if (evidence.reportedStatus === 'offline') return 'offline';
  if (now < evidence.leaseExpiresAtUnixMs) return 'online';
  return now < evidence.leaseExpiresAtUnixMs + agentLifecyclePolicy.reconnectGraceMs
    ? 'reconnecting'
    : 'offline';
}

export function evaluateAgentLifecycle(
  evidence: AgentPresenceEvidence,
  now: number,
  archiveAfterDays: number = agentLifecyclePolicy.archiveAfterDays,
): AgentLifecycle {
  const connection = agentConnection(evidence, now);
  const offlineSinceUnixMs =
    connection === 'offline'
      ? evidence.reportedStatus === 'offline'
        ? evidence.lastActiveAtUnixMs
        : evidence.leaseExpiresAtUnixMs + agentLifecyclePolicy.reconnectGraceMs
      : null;
  const archived =
    offlineSinceUnixMs !== null && now - offlineSinceUnixMs >= archiveAfterDays * 86_400_000;
  const listeningUntil = evidence.listeningUntilUnixMs;
  const shared = {
    offlineSinceUnixMs,
    archived,
    archiveReason: archived ? ('expired' as const) : null,
  };
  return connection === 'online'
    ? {
        ...shared,
        connection,
        reception:
          listeningUntil !== undefined && listeningUntil !== null && now < listeningUntil
            ? 'waiting'
            : listeningUntil === undefined
              ? 'unknown'
              : 'on_resume',
      }
    : { ...shared, connection, reception: 'unavailable' };
}

/** Apply capacity before searching or paging so clients cannot disagree by filter. */
export function projectAgentLifecycles(
  agents: readonly AgentPresenceEvidence[],
  now: number,
  archiveAfterDays: number = agentLifecyclePolicy.archiveAfterDays,
): ReadonlyMap<string, AgentLifecycle> {
  const result = new Map(
    agents.map((agent) => [agent.agentId, evaluateAgentLifecycle(agent, now, archiveAfterDays)]),
  );
  const offline = agents
    .filter((agent) => {
      const state = result.get(agent.agentId);
      return state?.connection === 'offline' && !state.archived;
    })
    .toSorted(
      (a, b) => b.lastActiveAtUnixMs - a.lastActiveAtUnixMs || a.agentId.localeCompare(b.agentId),
    );
  for (const agent of offline.slice(agentLifecyclePolicy.recentOfflineLimit)) {
    const state = result.get(agent.agentId);
    if (state) result.set(agent.agentId, { ...state, archived: true, archiveReason: 'capacity' });
  }
  return result;
}

File: crates/domain/src/agent_lifecycle.rs
//! Connection evidence and recoverable roster archival are independent of task completion.
use crate::agent_status::AgentWorkStatus;

pub const RECONNECT_GRACE_MS: i64 = 30_000;
pub const RECEPTION_FRESHNESS_MS: i64 = 15_000;
pub const DEFAULT_ARCHIVE_AFTER_DAYS: u16 = 7;
pub const RECENT_OFFLINE_LIMIT: usize = 100;
pub const PRESENCE_PAGE_SIZE: usize = 100;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AgentConnection {
    Online,
    Reconnecting,
    Offline,
}
impl AgentConnection {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Online => "online",
            Self::Reconnecting => "reconnecting",
            Self::Offline => "offline",
        }
    }
}
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AgentReception {
    Waiting,
    OnResume,
    Unknown,
    Unavailable,
}
impl AgentReception {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Waiting => "waiting",
            Self::OnResume => "on_resume",
            Self::Unknown => "unknown",
            Self::Unavailable => "unavailable",
        }
    }
}
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AgentArchiveReason {
    Expired,
    Capacity,
}
impl AgentArchiveReason {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Expired => "expired",
            Self::Capacity => "capacity",
        }
    }
}
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct AgentPresenceEvidence {
    pub reported_status: AgentWorkStatus,
    pub lease_expires_at: i64,
    pub last_active_at: i64,
    pub last_polled_at: Option<i64>,
    pub listening_until: Option<i64>,
    pub reception_known: bool,
}
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct AgentLifecycle {
    pub connection: AgentConnection,
    pub reception: AgentReception,
    pub offline_since: Option<i64>,
    pub archive_reason: Option<AgentArchiveReason>,
}
impl AgentPresenceEvidence {
    pub fn lifecycle(self, now: i64, archive_after_days: u16) -> AgentLifecycle {
        let connection = if self.reported_status == AgentWorkStatus::Offline {
            AgentConnection::Offline
        } else if now < self.lease_expires_at {
            AgentConnection::Online
        } else if now < self.lease_expires_at.saturating_add(RECONNECT_GRACE_MS) {
            AgentConnection::Reconnecting
        } else {
            AgentConnection::Offline
        };
        let offline_since = (connection == AgentConnection::Offline).then_some(
            if self.reported_status == AgentWorkStatus::Offline {
                self.last_active_at
            } else {
                self.lease_expires_at.saturating_add(RECONNECT_GRACE_MS)
            },
        );
        let reception = if connection != AgentConnection::Online {
            AgentReception::Unavailable
        } else if self.listening_until.is_some_and(|until| now < until) {
            AgentReception::Waiting
        } else if self.reception_known {
            AgentReception::OnResume
        } else {
            AgentReception::Unknown
        };
        AgentLifecycle {
            connection,
            reception,
            offline_since,
            archive_reason: offline_since
                .filter(|since| now - since >= i64::from(archive_after_days) * 86_400_000)
                .map(|_| AgentArchiveReason::Expired),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct AgentRosterPolicy {
    archive_after_days: u16,
}
impl AgentRosterPolicy {
    pub const fn new(days: u16) -> Option<Self> {
        if matches!(days, 1 | 7 | 30) {
            Some(Self {
                archive_after_days: days,
            })
        } else {
            None
        }
    }
    pub const fn archive_after_days(self) -> u16 {
        self.archive_after_days
    }
}
impl Default for AgentRosterPolicy {
    fn default() -> Self {
        Self {
            archive_after_days: DEFAULT_ARCHIVE_AFTER_DAYS,
        }
    }
}
